### PR TITLE
Queues: benchmark implementations against each other

### LINKED ISSUE #2221: Queues: benchmark specialized queues against binary heap variants
We have a small clutch of queue implementations now, and those have different capabilities. In increasing order of expressivity:

1. Round-Robin (resp. Strict) PIFOs can orchestrate `n` flows in a round-robin (resp. strict) policy. These are not strictly PIFOs: they cannot actually accept an item and a rank and enqueue that item with that rank.
2. Stable binary heaps can actually simulate PIFOs in a general sense. They take a value and a rank, and enqueue the value with that rank. Atop of stable binary heaps, we have built little layers of logic that grant ranks to values, such that {the layer of logic + the binary heap below it} together instruments a scheduling policy. We have such layers of logic for FIFO and PIFO policies. 

We have long had this vague notion that the simpler kinds of queues will be more performant. This issue seeks to test that for real. We will likely want to chat with @calebmkim, who has a little repository to measure LUTs, cycle counts, etc for a given design under a given set of inputs.

The first few concrete steps (just to get a comparison between 1 and 2 going) are:
- Implement two additional layers of logic on stable binary heaps:
    - `strict`
    - `round_robin`
- Check that these agree 100% with the outputs of the simplest kinds of queues (1, above).
- Measure the performance of each.

This work does not strictly have to happen during flag day, but I've put it in the flag day list because we will probably benefit from the tidying of flag day. Also, it kinda feels like this sort of benchmarking is "good housekeeping" when building multiple gadgets that have similar functionality.

### PR DESCRIPTION
**Pull request by polybeandip**
Closes #2221 

Changes:
- implement `strict` and `round_robin` scheduling transactions via `stable_binheap`
- use `generate_name` in `ComponentBuilder.case`; this way, there are less conflicts when a component uses multiple cases
    - change the associated `case` test
- tweak `gen_test_data.sh` to copy `round_robin` and `strict` data files into `tests/binheap/`
- make `clean_test_data.sh` purge data files in `tests/binheap/round_robin` and `tests/binheap/strict` 
- make `runt.toml` test our new queues
- setup `cycles.sh` to generate cycles counts for all tests
- setup `resources.sh` to generate synthesis results for all tests

### CODE REVIEW

**Comment by anshumanmohan** [comment]
Hi @polybeandip, thanks for getting this going! Since you are knee deep in this stuff, I would like to tack on a quick request that should be easy to handle. @parthsarkar17 is interested in adding some queue-ey stuff to his benchmarking suite that optimize FSMs and such. Could you please pass him designs for one or two of our more complicated tree schedulers? For each design, he'll need a .futil file, a .data file, and a .expect file.

Some candidates that come to mind: 
- Cassandra did up a "complex tree" that uses only `rr` and `strict` but has height of 2 or 3. 
- Some of your newest implementations of schedulers using stable binary heaps.

**Comment by anshumanmohan** [comment]
!! Please tell the channel!

**Comment by rachitnigam** [comment]
> It looks like @csziklai's PIFOs do about about 300,000 to 1,000,000 cycles better than the binary heaps!

This sounds awesome but I would caution against celebrating the numbers too much without synthesis results. Something that is takes 1,000,000 cycles at 100MHz is still way faster than something that take 100,000 cycles at 1MHz.

**Comment by polybeandip** [comment]
Agreed! Synthesis results are coming up shortly

**Comment by polybeandip** [comment]
**UPDATE** on our synthesis results

[`resources.sh`](https://github.com/calyxir/calyx/blob/2bd6dbe92b34780e230e2bab67992ecf566c5257/frontends/queues/resources.sh) generates the "executive summaries" of the synthesis reports: i.e. the output of 
```
fud e --to resource-estimate <filename>.futil
``` 
for all our designs. Notably, this means our results ([`resources.txt`](https://github.com/calyxir/calyx/blob/2bd6dbe92b34780e230e2bab67992ecf566c5257/frontends/queues/resources.txt)) are missing certain stats (for example, BRAMs).

@anshumanmohan I think it would be odd to commit `counts.txt` and `resources.txt` — I've only added them to show folks reviewing this PR the results. As for where they should live, I'm thinking we make a discussion thread in the [packet scheduling repo](https://github.com/cucapra/packet-scheduling) and park it there. Perhaps that thread could be a running log of synthesis results and cycles counts as we continue tweaking our designs.

**Comment by polybeandip** [comment]
Pretty pictures!

**Comment by rachitnigam** [comment]
> Notably, this means our results ([resources.txt](https://github.com/calyxir/calyx/blob/2bd6dbe92b34780e230e2bab67992ecf566c5257/frontends/queues/resources.txt)) are missing certain stats (for example, BRAMs) and use the default .xdc file: i.e. a clock period of 7 ns. Per @sampsyo's suggestion, this frequency should be okay for now.

Thanks for running synthesis @polybeandip! The frequency number you want to report is going to be determined using "worst_slack". You can calculate the best frequency using `best_freq = 1000/(7 - worst_slack) MHz`. In general, you want compare the product `cycle_count * best_freq`.

**Comment by polybeandip** [comment]
Ah, perfect! I'll crunch those numbers soon

**Comment by anshumanmohan** [comment]
> As for where they should live, I'm thinking we make a discussion thread in the [packet scheduling repo](https://github.com/cucapra/packet-scheduling) and park it there. Perhaps that thread could be a running log of synthesis results and cycles counts as we continue tweaking our designs?

Agreed, thank you!

**Comment by anshumanmohan** [comment]
@polybeandip so @ayakayorihiro has the same request as Parth! Would you mind sending her the same things?

File: frontends/queues/queues/binheap/flow_inference.py
Comment (anshumanmohan) [inline frontends/queues/queues/binheap/flow_inference.py:29]: Oh very snazzy! I'll use this pattern in some other places when I get around to a refactor.

File: frontends/queues/queues/binheap/round_robin.py
Comment (anshumanmohan) [inline frontends/queues/queues/binheap/round_robin.py:91]: Very minor, but consider pulling out into its own list?

File: frontends/queues/queues/binheap/strict.py
Comment (anshumanmohan) [inline frontends/queues/queues/binheap/strict.py:46]: Very clean!

File: frontends/queues/queues/strict_or_rr.py
Comment (anshumanmohan) [inline frontends/queues/queues/strict_or_rr.py:128]: Interesting, so how was this working out before?

File: frontends/queues/resources.txt
Comment (anshumanmohan) [inline frontends/queues/resources.txt]: Did we discuss that this file can just truly become a json?

**Review by anshumanmohan — CHANGES_REQUESTED** [review]
Very cool stuff, and thanks for leading our foray into real benchmarking! I have a couple notes, one of which I will seriously request, but otherwise I have reviewed to learn!

File: frontends/queues/resources.txt
Comment (polybeandip) [reply frontends/queues/resources.txt]: Yep, it totally can and same with `cycles.txt`, not making them jsons was pretty silly of me.

I'll tweak `cycles.sh` and `resources.sh` to do that.

File: frontends/queues/queues/strict_or_rr.py
Comment (polybeandip) [reply frontends/queues/queues/strict_or_rr.py:128]: This is just a silly python thing.

If you throw a list into a function that takes a variable number of args, it's not that the function is called with list entries as arguments. Instead, the function is called with one argument: the entire list.

For example, try running this:
```
def f(*args):
    for arg in args:
        print(arg)

print("calling f([1, 2, 3])")
f([1,2,3])
print("calling f(*[1, 2, 3])")
f(*[1,2,3])
```

For our code, this means we were throwing a list (i.e. a `seq` block) into `cb.par`. So, the resulting `futil` code would look like 
```
par { 
    seq {
        ... # guarded subqueue invokes
    }
} 
```
instead of 
```
par { 
     ... # guarded subqueue invokes
} 
```
Both are fine for correctness, but I think we're accidentally not running them in parallel.

Comment (anshumanmohan) [reply frontends/queues/queues/strict_or_rr.py:128]: Ah gotcha, I think I have caught something a little like this before but I needed my memory jogged! Probably a good thing for us to keep in mind as we try to fine-tune our implementations. This sort of thing is money left on the table!

File: calyx-py/calyx/builder.py
Comment (polybeandip) [reply calyx-py/calyx/builder.py:224]: > the same two things need to be compared repeatedly?

Admittedly, I'm having a hard time parsing this. Which two things are compared repeatedly?

The reason for this tweak was to allow the user to have multiple case statements depending on the same signal; their bodies can be totally different though. This seems like a reasonable use-case to me?

For example, in `binheap/round_robin.py` we have two case statements depending on `flow_in.out` (i.e. the flow of our incoming packet):
```
comp.case(flow_in.out, binheap_invokes)
```
so that we push to the heap with the correct rank and
```
comp.case(flow_in.out, rank_ptr_incrs)
```
so that we update our state correctly by incrementing the right rank pointer.

Previously, the eDSL would complain both calls to `comp.case` instantiate comparators called `flow_in_out_eq_<branch number>`.

File: frontends/queues/queues/binheap/round_robin.py
Comment (polybeandip) [reply frontends/queues/queues/binheap/round_robin.py]: Done via [5a0c162](https://github.com/calyxir/calyx/pull/2276/commits/5a0c1627e53eef8a54fb211f076e6a57d003335d)

File: frontends/queues/resources.txt
Comment (polybeandip) [reply frontends/queues/resources.txt]: Done via [5eff826](https://github.com/calyxir/calyx/pull/2276/commits/5eff8265450afa0682caa0b60720e0a5b4ba8cbe)

**Comment by anshumanmohan** [comment]
Terrific! So this is ready to merge, once you refactor the underlying issue and link this PR to the underlying issue. Go for it, no need for another review from me

**Comment by polybeandip** [comment]
Okay! I'm still a little unsure if I have the green light to make the tweak I did to `case` statements in `Calyx-Py`. Did you have a chance to look at [this comment](https://github.com/calyxir/calyx/pull/2276#discussion_r1773971113)?

**Comment by anshumanmohan** [comment]
Yeah! I replied just below. The comment starting with “ I understand now. For completeness…”. Green light!

**Pull request merged**
